service/index/writer.go

```diff
@@ -150,3 +169,40 @@ func (w *Writer) Events(height uint64, events []flow.Event) error {
 	}
 	return nil
 }
+
+func (w *Writer) apply(op func(*badger.Txn) error) error {
+	w.RLock()
+	err := w.err
+	w.RUnlock()
+	if err != nil {
+		return fmt.Errorf("could not commit transaction: %w", w.err)
+	}
+	err = op(w.tx)
+	if errors.Is(err, badger.ErrTxnTooBig) {
+		w.tx.CommitWith(w.done)
+		_ = w.sema.Acquire(context.Background(), 1)
+		w.tx = w.db.NewTransaction(true)
+		err = op(w.tx)
+	}
+	if err != nil {
+		return fmt.Errorf("could not apply operation: %w", err)
+	}
+	return nil
+}
+
+func (w *Writer) done(err error) {
+	if err != nil {
+		w.Lock()
+		w.err = err
+		w.Unlock()
+	}
+	w.sema.Release(1)
+}
+
+func (w *Writer) Close() error {
+	_ = w.sema.Acquire(context.Background(), int64(w.cfg.ConcurrentTransactions))
+	if w.err != nil {
+		return fmt.Errorf("could not flush transactions: %w", w.err)
+	}
+	return nil
+}
```

Comment:
I find this part slightly unreadable.

WDYT of tweaking it like this? It's not much but I feel like it's clearer this way

```suggestion
func (w *Writer) error() error {
	w.RLock()
	defer w.RUnlock()
	err := w.err
	return err
}

func (w *Writer) setError(err error) {
	w.RLock()
	defer w.RUnlock()
	w.err = err
}

func (w *Writer) apply(op func(*badger.Txn) error) error {
	if w.error() != nil {
		return fmt.Errorf("could not commit transaction: %w", w.err)
	}
	
	err := op(w.tx)
	if errors.Is(err, badger.ErrTxnTooBig) {
		w.tx.CommitWith(w.done)
		_ = w.sema.Acquire(context.Background(), 1)
		w.tx = w.db.NewTransaction(true)
		err = op(w.tx)
	}
	if err != nil {
		return fmt.Errorf("could not apply operation: %w", err)
	}
	return nil
}

func (w *Writer) done(err error) {
	if err != nil {
		w.setError(err)
	}
	w.sema.Release(1)
}

func (w *Writer) Close() error {
	_ = w.sema.Acquire(context.Background(), int64(w.cfg.ConcurrentTransactions))
	if w.err != nil {
		return fmt.Errorf("could not flush transactions: %w", w.err)
	}
	return nil
}
```

Reply:
To be honest, I don't think your suggestion increases readability. It's just extra function bodies for little value. I think just putting some code comments around the blocks would make more sense.